Mathlib/Data/QPF/Multivariate/Constructions/Fix.lean

```diff
@@ -0,0 +1,370 @@
+
+ * `Fix.mk`     - constructor
+ * `Fix.dest`    - destructor
+ * `Fix.rec`    - recursor: basis for defining functions by structural recursion on `Fix F α`
+ * `Fix.drec`   - dependent recursor: generalization of `Fix.rec` where
+                  the result type of the function is allowed to depend on the `Fix F α` value
+ * `Fix.rec_eq` - defining equation for `recursor`
+ * `Fix.ind`    - induction principle for `Fix F α`
+
+## Implementation notes
+
+For `F` a `QPF`, we define `Fix F α` in terms of the W-type of the polynomial functor `P` of `F`.
+We define the relation `WEquiv` and take its quotient as the definition of `Fix F α`.
+
+See [avigad-carneiro-hudon2019] for more details.
+
+## Reference
+
+ * Jeremy Avigad, Mario M. Carneiro and Simon Hudon.
+   [*Data Types as Quotients of Polynomial Functors*][avigad-carneiro-hudon2019]
+-/
+
+
+universe u v
+
+namespace MvQPF
+
+open TypeVec
+
+open MvFunctor (LiftP LiftR)
+
+open MvFunctor
+
+variable {n : ℕ} {F : TypeVec.{u} (n + 1) → Type u} [MvFunctor F] [q : MvQPF F]
+
+
+/-- `recF` is used as a basis for defining the recursor on `Fix F α`. `recF`
+traverses recursively the W-type generated by `q.P` using a function on `F`
+as a recursive step -/
+def recF {α : TypeVec n} {β : Type _} (g : F (α.append1 β) → β) : q.P.W α → β :=
+  q.P.wRec fun a f' _f rec => g (abs ⟨a, splitFun f' rec⟩)
+set_option linter.uppercaseLean3 false in
+#align mvqpf.recF MvQPF.recF
+
+theorem recF_eq {α : TypeVec n} {β : Type _} (g : F (α.append1 β) → β) (a : q.P.A)
+    (f' : q.P.drop.B a ⟹ α) (f : q.P.last.B a → q.P.W α) :
+    recF g (q.P.wMk a f' f) = g (abs ⟨a, splitFun f' (recF g ∘ f)⟩) := by
+  rw [recF, MvPFunctor.wRec_eq]; rfl
+set_option linter.uppercaseLean3 false in
+#align mvqpf.recF_eq MvQPF.recF_eq
+
+theorem recF_eq' {α : TypeVec n} {β : Type _} (g : F (α.append1 β) → β) (x : q.P.W α) :
+    recF g x = g (abs (appendFun id (recF g) <$$> q.P.wDest' x)) := by
+  apply q.P.w_cases _ x
+  intro a f' f
+  rw [recF_eq, q.P.wDest'_wMk, MvPFunctor.map_eq, appendFun_comp_splitFun, TypeVec.id_comp]
+set_option linter.uppercaseLean3 false in
+#align mvqpf.recF_eq' MvQPF.recF_eq'
+
+/-- Equivalence relation on W-types that represent the same `Fix F`
+value -/
+inductive WEquiv {α : TypeVec n} : q.P.W α → q.P.W α → Prop
+  | ind (a : q.P.A) (f' : q.P.drop.B a ⟹ α) (f₀ f₁ : q.P.last.B a → q.P.W α) :
+    (∀ x, WEquiv (f₀ x) (f₁ x)) → WEquiv (q.P.wMk a f' f₀) (q.P.wMk a f' f₁)
+  | abs (a₀ : q.P.A) (f'₀ : q.P.drop.B a₀ ⟹ α) (f₀ : q.P.last.B a₀ → q.P.W α) (a₁ : q.P.A)
+    (f'₁ : q.P.drop.B a₁ ⟹ α) (f₁ : q.P.last.B a₁ → q.P.W α) :
+    abs ⟨a₀, q.P.appendContents f'₀ f₀⟩ = abs ⟨a₁, q.P.appendContents f'₁ f₁⟩ →
+      WEquiv (q.P.wMk a₀ f'₀ f₀) (q.P.wMk a₁ f'₁ f₁)
+  | trans (u v w : q.P.W α) : WEquiv u v → WEquiv v w → WEquiv u w
+set_option linter.uppercaseLean3 false in
+#align mvqpf.Wequiv MvQPF.WEquiv
+
+theorem recF_eq_of_wequiv (α : TypeVec n) {β : Type _} (u : F (α.append1 β) → β) (x y : q.P.W α) :
```

Comment:
```suggestion
theorem recF_eq_of_wEquiv (α : TypeVec n) {β : Type _} (u : F (α.append1 β) → β) (x y : q.P.W α) :
```

Reply:
I think `of_WEquiv` is better here and elsewhere actually. The main type is `WEquiv`

Reply:
The [https://github.com/leanprover-community/mathlib4/wiki/Porting-wiki#naming-convention](convention) says:
> When something named with UpperCamelCase is part of something named with snake_case, it is referenced in lowerCamelCase

So, indeed the main type is `WEquiv`, but we spell it `wEquiv` when it's part of a snake_case theorem. (Admittedly, I don't think we are being very consistent with this part of the convention)

Reply:
you're correct... `WEquiv` is `FooEquiv` which would be `of_fooEquiv`